napari/components/layerlist.py

```diff
@@ -126,8 +126,46 @@ def _update_name(self, event):
         layer = event.source
         layer.name = self._coerce_name(layer.name, layer)
 
+    def __setitem__(self, key, value):
+        # we have to repeat some logic even though it happens in the superclass
+        # because we need to ensure inplace changes such as reshuffling layers still work
+        # e.g: layers[2:5] = sorted(layers[2:5])
+        old = self._list[key]
+        if isinstance(key, slice):
+            if not isinstance(value, Iterable):
+                raise TypeError(
+                    trans._(
+                        'Can only assign an iterable to slice',
+                        deferred=True,
+                    )
+                )
+            # ensure uniqueness
+            if len(value) != len(set(value)):
+                raise ValueError(
+                    trans._("LayerList cannot contain duplicate layers.")
+                )
+            for v in value:
+                # here we use `not in` rather than checking element by element,
+                # because we allow reshuffling!
+                if v in self._list and v not in old:
+                    raise ValueError(
+                        f"Layer '{v}' is already present in layer list"
+                    )
+        elif isinstance(key, int):
+            if value in self._list and value is not old:
```

Comment:
since this is repeated in three places, why not extract it?:

```python
def _ensure_unique(self, values, allow=()):
    bad = set(self._list) - set(allow)
    for v in (values if isinstance(values, tuple) else (values,)):
        if v in bad:
            raise ValueError(
                f"Layer '{v}' is already present in layer list"
            )
```

and then when you have a single `value`, just use `_ensure_unique((value,), (old,))`